### PR TITLE
Inject set of ConfigurationInitializers to HdfsConfigurationInitializer

### PR DESCRIPTION
**Pull request by kokosing**
Inject set of ConfigurationInitializers to HdfsConfigurationInitializer

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HdfsConfigurationInitializer.java [hunk 1]
```diff
@@ -92,8 +92,7 @@ public HdfsConfigurationInitializer(HiveClientConfig config, S3ConfigurationUpda
         this.isHdfsWireEncryptionEnabled = config.isHdfsWireEncryptionEnabled();
         this.textMaxLineLength = toIntExact(config.getTextMaxLineLength().toBytes());
 
-        this.s3ConfigurationUpdater = requireNonNull(s3ConfigurationUpdater, "s3ConfigurationUpdater is null");
-        this.gcsConfigurationInitialize = requireNonNull(gcsConfigurationInitialize, "gcsConfigurationInitialize is null");
+        this.configurationInitializers = ImmutableSet.copyOf(requireNonNull(configurationInitializers, "configurationUpdaters is null"));
```
Comment (luohao) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HdfsConfigurationInitializer.java]: > "configurationUpdaters is null"

--> "configurationInitializers is null"

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HdfsConfigurationInitializer.java [hunk 2]
```diff
@@ -146,8 +145,7 @@ public void initializeConfiguration(Configuration config)
 
         configureCompression(config, compressionCodec);
 
-        s3ConfigurationUpdater.updateConfiguration(config);
-        gcsConfigurationInitialize.updateConfiguration(config);
+        configurationInitializers.forEach(configurationUpdater -> configurationUpdater.initializeConfiguration(config));
```
Comment (luohao) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HdfsConfigurationInitializer.java]: should we rename `configurationUpdater` to `configurationInitializer`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3ConfigurationUpdater.java [hunk 3]
```diff
@@ -21,8 +21,34 @@
...
+import static io.prestosql.plugin.hive.s3.PrestoS3FileSystem.S3_SSE_KMS_KEY_ID;
+import static io.prestosql.plugin.hive.s3.PrestoS3FileSystem.S3_SSE_TYPE;
+import static io.prestosql.plugin.hive.s3.PrestoS3FileSystem.S3_SSL_ENABLED;
+import static io.prestosql.plugin.hive.s3.PrestoS3FileSystem.S3_STAGING_DIRECTORY;
+import static io.prestosql.plugin.hive.s3.PrestoS3FileSystem.S3_USER_AGENT_PREFIX;
+import static io.prestosql.plugin.hive.s3.PrestoS3FileSystem.S3_USE_INSTANCE_CREDENTIALS;
+
 public class PrestoS3ConfigurationUpdater
-        implements S3ConfigurationUpdater
+        implements ConfigurationInitializer
```
Comment (luohao) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3ConfigurationUpdater.java]: Should we also rename  `PrestoS3ConfigurationUpdater` to `PrestoS3ConfigurationInitializer`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java [hunk 4]
```diff
@@ -61,10 +62,10 @@ private static void validateEmrFsClass()
     }
 
     public static class EmrFsS3ConfigurationUpdater
```
Comment (luohao) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java]: Should we rename it to `EmrFsS3ConfigurationInitializer`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/ConfigurationInitializer.java [hunk 5]
```diff
@@ -15,7 +15,7 @@
 
 import org.apache.hadoop.conf.Configuration;
 
-public interface S3ConfigurationUpdater
+public interface ConfigurationInitializer
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/ConfigurationInitializer.java]: How is `ConfiguratonInitializer` similar to `DynamicConfigurationProvider`?
Could we just use `DynamicConfigurationProvider`?

Probably we should rename `DynamicConfigurationProvider` to `DynamicConfigurationUpdater` if it's still needed.

Comment (kokosing) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/s3/ConfigurationInitializer.java]: Initializer is used only once, while `DynamicConfigurationProvider` is used to provide session based configuration (which is based on the one that was previously initialized).

**Comment by kokosing** [comment]
Thank you for review, comments are addressed.

**Comment by sopel39** [comment]
LGTM % @luohao

**Comment by luohao** [comment]
The first 3 commits look good to me. There are failures in unit tests related to constructor argument, should be an easy fix.

Regarding the last commit, we had the same discussion about the naming in my PR. I am OK with either `*Updater` or `*Provider` (I am very bad at picking good names). @electrum suggests using `*Provider`, so I would like to hear his thoughts on this.

**Comment by kokosing** [comment]
> Regarding the last commit

I will extract it to separate pull request, there is no point to block this PR because of this.

**Comment by luohao** [comment]
> > Regarding the last commit
> 
> I will extract it to separate pull request, there is no point to block this PR because of this.

Otherwise, it looks good to me.

**Pull request merged**